lib/logging/CLogger.cpp

```diff
@@ -1,404 +1,408 @@
-{
-	CLoggerDomain currentDomain = domain;
-	while(true)
-	{
-		const auto & loggerPair = map.find(currentDomain.getName());
-		if(loggerPair != map.end())
-		{
-			const auto & levelMap = loggerPair->second;
-			const auto & levelPair = levelMap.find(level);
-			if(levelPair != levelMap.end())
-				return levelPair->second;
-		}
-
-		if (currentDomain.isGlobalDomain())
-			break;
-
-		currentDomain = currentDomain.getParent();
-	}
-
-	throw std::runtime_error("failed to find color for requested domain/level pair");
-}
-
-CLogConsoleTarget::CLogConsoleTarget(CConsoleHandler * console) : console(console), threshold(ELogLevel::INFO), coloredOutputEnabled(true)
-{
-	formatter.setPattern("%m");
-}
-
-void CLogConsoleTarget::write(const LogRecord & record)
-{
-	if(threshold > record.level)
-		return;
-
-	std::string message = formatter.format(record);
-
-#ifdef VCMI_ANDROID
-    __android_log_write(ELogLevel::toAndroid(record.level), ("VCMI-" + record.domain.getName()).c_str(), message.c_str());
-#else
-
-	const bool printToStdErr = record.level >= ELogLevel::WARN;
-	if(console)
-	{
-		const EConsoleTextColor::EConsoleTextColor textColor =
-			coloredOutputEnabled ? colorMapping.getColorFor(record.domain, record.level) : EConsoleTextColor::DEFAULT;
-
-		console->print(message, true, textColor, printToStdErr);
-	}
-	else
-	{
-		TLockGuard _(mx);
-		if(printToStdErr)
-			std::cerr << message << std::endl;
-		else
-			std::cout << message << std::endl;
-	}
-#endif
-}
-
-bool CLogConsoleTarget::isColoredOutputEnabled() const { return coloredOutputEnabled; }
-void CLogConsoleTarget::setColoredOutputEnabled(bool coloredOutputEnabled) { this->coloredOutputEnabled = coloredOutputEnabled; }
-
-ELogLevel::ELogLevel CLogConsoleTarget::getThreshold() const { return threshold; }
-void CLogConsoleTarget::setThreshold(ELogLevel::ELogLevel threshold) { this->threshold = threshold; }
-
-const CLogFormatter & CLogConsoleTarget::getFormatter() const { return formatter; }
-void CLogConsoleTarget::setFormatter(const CLogFormatter & formatter) { this->formatter = formatter; }
-
-const CColorMapping & CLogConsoleTarget::getColorMapping() const { return colorMapping; }
-void CLogConsoleTarget::setColorMapping(const CColorMapping & colorMapping) { this->colorMapping = colorMapping; }
-
-CLogFileTarget::CLogFileTarget(boost::filesystem::path filePath, bool append)
-	: file(std::move(filePath), append ? std::ios_base::app : std::ios_base::out)
-{
-//	formatter.setPattern("%d %l %n [%t] - %m");
-	formatter.setPattern("%l %n [%t] - %m");
-}
-
-void CLogFileTarget::write(const LogRecord & record)
-{
-	std::string message = formatter.format(record); //formatting is slow, do it outside the lock
-	TLockGuard _(mx);
-	file << message << std::endl;
-}
-
-const CLogFormatter & CLogFileTarget::getFormatter() const { return formatter; }
-void CLogFileTarget::setFormatter(const CLogFormatter & formatter) { this->formatter = formatter; }
+/*

```

Comment:
Possibly line ending issue, please fix.

Reply:
> Not bad for now, except minor style issues.
> 
> However, if you're going for more complex AI, you might want to try fuzzy logic to prioritize actions. Here it will work better than in adventure AI since the number of possible actions is lower.

We'll see. So far my primary concern was fixing some really obvious flaws in the AI, writing the necessary utility functions and laying the groundwork for future improvements.